### PR TITLE
feat(dotnet): Add .NET Core support

### PR DESCRIPTION
**Pull request by omajid**
This adds support for generating simple ASP.NET Core-based REST endpoints and database-using CRUD-style REST endpoints.

The following new capabilities are supported:
- C# language
- .NET Core platform
 - REST API endpoint using ASP.NET Core
  - CRUD/REST API endpoint using ASP.NET Core/EF Core

Both MySQL and PostgreSQL are supported for CRUD.

Only .NET Core version 2.2, the latest "Current" release, is supported.

This uses the RHEL-based .NET Core s2i images. The sources of those images are also accessible via https://github.com/redhat-developer/s2i-dotnetcore

The .NET Core icon is taken from https://github.com/openshift/origin-web-console/blob/master/app/images/logos/dotnet.svg

Open questions:
1. ~~How can I add automated tests for this? I would to make sure this continues working if anything changes in the backend here or in the s2i container images~~
2. The architecture here is a bit weird. The C# language is the base. The .NET Platform derives from it. And ASP.NET Core is built on top. In reality, C#, F# and VB.NET can all run on top of the .NET Platform. (It's not very different than scala, kotlin and java all running on top of the JVM). I haven't been able to express this in code.
3. ~~I see that the official Red Hat image for OpenJDK is used. Can I use the official .NET Core images from Red Hat too?~~
4. Where do we go from here? Will this be merged/shipped (for preview) right now? Or should I look into adding additional capabilities/generators? Is a reviewer going to try out and okay the end-to-end experience?

### CODE REVIEW

File: catalog/generators/rest-aspnetcore/index.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        if (!await this.filesCopied()) {
+            // First copy the files from the base DoNet platform module
+            // and then copy our own over that
+            const pprops = {
+                'application': props.application,
+                'subFolderName': props.subFolderName,
+                'serviceName': props.serviceName,
+                'routeName': props.routeName,
+                'env': {
+                    'DOTNET_SDK_VERSION': '2.1.500'
```
Comment (omajid) [inline catalog/generators/rest-aspnetcore/index.ts]: This is a workaround for https://github.com/redhat-developer/s2i-dotnetcore/issues/263

File: catalog/generators/rest-aspnetcore/info.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "type": "generator",
+  "name": "ASP.NET COre REST",
```
Comment (gastaldi) [inline catalog/generators/rest-aspnetcore/info.json]:
```suggestion
  "name": "ASP.NET Core REST",
```

Comment (omajid) [reply catalog/generators/rest-aspnetcore/info.json]: Thanks. Fixed.

**Comment by ALRubinger** [comment]
Yessss!  Cannot wait to try my hand at some .NET!

**Comment by quintesse** [comment]
@omajid Very nice work!

To answer some of your questions:

> How can I add automated tests for this? I would to make sure this continues working if anything changes in the backend here or in the s2i container images

Running `yarn ittest` should test all the capabilities in all the runtimes. If you just want to test yours you can use `yarn ittest --runtimes=aspnetcore`

> The architecture here is a bit weird. The C# language is the base. The .NET Platform derives from it. And ASP.NET Core is built on top. In reality, C#, F# and VB.NET can all run on top of the .NET Platform. (It's not very different than scala, kotlin and java all running on top of the JVM). I haven't been able to express this in code.

Yes, you're right that the naming isn't ideal and it's likely that what now is called "language" will become "platform" (with "language-java" becoming "platform-jvm") and "platform" becoming "runtime" (with "platform-vertx" for example becoming "runtime-vertx"). If you want you can change "language-csharp" to "language-dotnet" and "platform-dotnet" to "platform-aspnet" or something like that. I'll leave it up to you.

> I see that the official Red Hat image for OpenJDK is used. Can I use the official .NET Core images from Red Hat too?

I don't know the answer to that, but I would expect so.

> Where do we go from here? Will this be merged/shipped (for preview) right now? Or should I look into adding additional capabilities/generators? Is a reviewer going to try out and okay the end-to-end experience?

For merging I'd ask you to at least also implement the Database capability (with its support for both MySql and PostgreSQL) and then I'd be all for merging it. We'll definitely try it out ASAP! But if it has the two capabilities and passes the integration tests I don't see any obstacle to merging and shipping this.

File: lib/core/catalog/types.ts [hunk 3]
```diff
@@ -118,6 +118,16 @@ abstract class BaseCatalogItem implements CatalogItem {
         });
     }
 
+    protected move(original: string, to: string, options?: MoveOptions): Promise<void> {
```
Comment (quintesse) [inline lib/core/catalog/types.ts]: It's nice to have this option but I'd prefer these functions to be very specific in their use and have as little options and arguments as possible. So could we perhaps remove the options and just always use `overwrite: true`?

**Comment by ALRubinger** [comment]
>> I see that the official Red Hat image for OpenJDK is used. Can I use the official .NET Core images from Red Hat too?

>I don't know the answer to that, but I would expect so.

This has been this historical preference of PMs for our runtimes thus far, so it's my suggestion as well.  Ping Bob Davis for confirmation if he'd like to make a business case to use a community .NET image instead. :)

**Comment by tmds** [comment]
We have different .NET Core images for different .NET Core version. The example code is now assuming the 2.1 version is being used. We should think about how we want to deal with those versions. In particular, to minimize work for major and patch releases.

From UX point of view, how does the user pick the version? Is this by selecting an image?

Can we derive the version from the image the user has selected?
Based on that, we may pick a different template (e.g. using a versioned folder).
Or maybe, we can make our example simple enough, so it works across versions, we can fill in the `TargetFramework` in the csproj file with the appropriate `netcoreappx.y` version when the project is generated.

**Comment by quintesse** [comment]
@tmds right now no version selection of platforms or runtimes is supported. This is something that is high on our TODO list but for now I'd limit the support to just a single version. Just choose the one that you people think is best (the newest, the most popular, etc)

We'll come up with a way to manage different versions of the generator files but I'd suggest looking at way too minimize the differences between versions to make maintenance easier. Perhaps you can get away with simply adjusting some configuration files. (The code has several useful functions for patching files)

**Comment by quintesse** [comment]
@omajid there's another thing that needs to be changed: the detection of whether a codebase uses .net or not: https://github.com/fabric8-launcher/launcher-creator-backend/blob/master/lib/core/analysis/index.ts#L26

It's just a function that given a directory with code returns the builder image that should be used.

**Comment by tmds** [comment]
> This is something that is high on our TODO list but for now I'd limit the support to just a single version.

Multiple versions is really a thing in .NET Core.
There are 1 to 2 new versions per year, and users expect to be able to pick the latest version.
The latest version can be retired already 3 months after a new release (no longer receiving security patches).

**Comment by ALRubinger** [comment]
@tmds Cool, thanks for that context.  It helps us prioritize feature development.

To get out the application creator feature we scoped down, and version selection was one of the concessions we made in the first couple iterations.  Definitely we'll be slating support for N versions in the runtime selections, and by adding the .NET perspective here you make a good case for accelerating that.

My recommendation now is to treat this like tech preview and offer users the latest runtime available when creating new applications.  When we drop in multiversion support here, we can then let ya scale it out to older ones. 

Sound like a good plan?

File: lib/core/catalog/types.ts (hunk 3 above)
Comment (omajid) [reply lib/core/catalog/types.ts]: Fixed now.

**Comment by omajid** [comment]
> Running yarn ittest should test all the capabilities in all the runtimes. If you just want to test yours you can use yarn ittest --runtimes=aspnetcore

That doesn't work on my end:

```
$ yarn ittest --runtime=dotnet
yarn run v1.13.0
$ NODE_PATH=lib:catalog mocha --opts it-test/mocha.opts -r ts-node/register it-test/all-tests.ts --runtime=dotnet


  1) "before all" hook
  2) "after all" hook

  0 passing (7s)
  2 failing

  1) "before all" hook:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  

  2) "after all" hook:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  



error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Any ideas on how to fix this? I expect that my implementations are like 95% of what the expectations from REST/CRUD are. I would like it to hit 100%.

>  If you want you can change "language-csharp" to "language-dotnet" and "platform-dotnet" to "platform-aspnet" or something like that. I'll leave it up to you.

I have left that as it was, with the exception of just using `rest-dotnet` and `crud-dotnet` instead of the more specific names (`rest-aspnetcore` and `crud-aspnetcore+efcore`).

> For merging I'd ask you to at least also implement the Database capability (with its support for both MySql and PostgreSQL) and then I'd be all for merging it. 

Done

> My recommendation now is to treat this like tech preview and offer users the latest runtime available when creating new applications.

I have implemented the latest version that has a Long Term Support (2.1).

**Comment by gastaldi** [comment]
@omajid the correct command is `yarn ittest --runtimes=dotnet` (runtimes, not runtime)

**Comment by omajid** [comment]
Still no luck :(

```
$ yarn ittest --runtimes=dotnet
yarn run v1.13.0
$ NODE_PATH=lib:catalog mocha --opts it-test/mocha.opts -r ts-node/register it-test/all-tests.ts --runtimes=dotnet


  1) "before all" hook
  2) "after all" hook

  0 passing (6s)
  2 failing

  1) "before all" hook:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  

  2) "after all" hook:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  



error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by gastaldi** [comment]
@omajid try running 
```
yarn ittest --runtimes=dotnet --verbose --timeout 10000
```

**Comment by quintesse** [comment]
@omajid lastthing to change, promise (I hope)! https://github.com/fabric8-launcher/launcher-creator-backend/blob/master/catalog/generators/import-codebase/index.ts#L15 , please add an if for dotnet here.

I need to document this ASAP, I just didn't expect so many new languages so soon! (Which is great)

File: catalog/generators/database-crud-dotnet/index.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                    },
+                    'DB_PASSWORD': {
+                        'secret': props.secretName,
+                        'key': 'password'
+                    }
+                }
+            } as PlatformDotNetProps;
+            // First copy the files from the base .NET Core platform module
+            // and then copy our own over that
+            console.log(props);
```
Comment (quintesse) [inline catalog/generators/database-crud-dotnet/index.ts]: Don't forget to remove this :-)

Comment (omajid) [reply catalog/generators/database-crud-dotnet/index.ts]: Sorry, leftover. Fixed now.

**Comment by quintesse** [comment]
@omajid so were you able to run the tests this time? Do they all pass? :-)

**Comment by quintesse** [comment]
Ok, although if you can get it to run against minishift I'll already be happy

**Comment by tmds** [comment]
> My recommendation now is to treat this like tech preview and offer users the latest runtime available when creating new applications. When we drop in multiversion support here, we can then let ya scale it out to older ones.
Sound like a good plan?

I think we may set a wrong expectation if we add support for this when not knowing how to deal with multiple versions. Dealing with several versions is important to understand the cost in maintaining this.

Do you have an idea when multiple runtime version support fits in your planning?

From a user perspective, there should be at least 2 versions. The latest LTS and the latest Current release. The latest LTS is 2.1, which is 1 year old. The latest current is 2.2, which was released Dec 2018. At about the same time, a preview release of 3.0 was made available. In second half of 2019, 3.0 will be released. Most likely, this will be a Current release. So if you only have 2.1, you seem to be lagging.

If we want to pick a single version for 'tech preview', maybe we should pick the latest? That won't make it look as if we are lagging. And we'll be confronted more with the actual maintenance.

**Comment by ALRubinger** [comment]
@tmds If this is important to you, it's important to us. :)  We aim for this system to fit the needs of the runtime teams to best reach their target users.

I've opened https://github.com/fabric8-launcher/launcher-planning/issues/210 to track support for multiple versions in the create flow.  Will leave it to the team to estimate delivery.  @quintesse, wanna have a look and get back to @tmds on this?  For after the L-Train we're cutting this week. :)

> If we want to pick a single version for 'tech preview', maybe we should pick the latest?

Yep, that's what I'd do.  I'd grab the most cutting edge thing with no support offering and make it available, then add LTS and other versions as we go.  I'm all about incrementally adding value and wouldn't block people from trying the 3.0 preview until we can also give them options for 2.2 or LTS 2.1.  Or choose 2.2 first, your pick! :P

**Comment by quintesse** [comment]
@tmds adding that feature shouldn't take too long, the backend basically already supports versions, we just need to come up with a design for the frontend and test it all works.

The maintenance burden for the runtime teams depends on their particular cases but really only depends on the amount of changes that would be needed to adjust a code base for version X to one for version Y. The more similar they are the easier it will be to maintain. If they only difference is a version number in a build file for example then the solution is a one-liner.

**Comment by omajid** [comment]
I tried running tests, without any of my changes and I still can't get them to pass. Any ideas? Here's what I did: https://gist.github.com/omajid/921250ad824089dffa23d27c56433ddb

**Comment by gastaldi** [comment]
@omajid try running the tests for the dotnet runtime only: `yarn ittest --runtimes=dotnet --verbose --timeout=10000`

**Comment by omajid** [comment]
I just ran `$ yarn ittest --runtimes=dotnet --verbose --timeout 1000000` and the output still includes lines like: 

    Runtime nodejs with database,health,rest,welcome

:confused:

**Comment by quintesse** [comment]
Btw, there should be no need to add a timeout anymore because it now gets set to 10s by default

**Comment by quintesse** [comment]
Also, if you update to the latest code from master the `--verbose` option will now show much more information about any commands that might be failing to run

**Comment by omajid** [comment]
I have a bit of a dilemma. I was able to run tests and get failures. The database tests are failing because .NET Core really likes to return HTTP 400 (bad request) errors instead of 415 (unsupported media type) or 422 (unprocessible entity). Is it worth complicating the code significantly to make sure the correct 4xx error codes are returned by the rest endpoints?

**Comment by gastaldi** [comment]
@omajid IMHO following the API contract is a key point here, so +1 to handle errors and return the expected code in the tests

**Comment by quintesse** [comment]
> IMHO following the API contract is a key point here, so +1 to handle errors and return the expected code in the tests

@gastaldi I don't necessarily disagree with you, but we have the same problem with Thorntail and WIldfly (and in both cases these still aren't fixed) which seems to suggest to me that this is a pretty common situation (where the framework handles part of the errors).

And making the code more complex perhaps makes it less of a nice example of how to do things for that particular framework/runtime.

So although I'm all for implementing the API strictly for all correct/non-failure/200 requests I'm wondering if we should perhaps be more lenient with errors? I'm not saying that we should accept a 500 when a 404 was expected, but perhaps we could treat things like 400, 415, 422 in a similar way?

**Comment by tmds** [comment]
> I'm wondering if we should perhaps take a pragmatic approach and be more lenient with errors?

+1 for idiomatic usage of the framework. Also this reduces custom changes needed for each new version.

**Comment by tmds** [comment]
> @tmds adding that feature shouldn't take too long, the backend basically already supports versions, we just need to come up with a design for the frontend and test it all works.

It's not clear for me how the backend supports multiple versions, each runtime seems to deal in its own way with the images. Anyway, good that it won't take too long. And we'll try to use it when the feature lands.

**Comment by tmds** [comment]
> If we want to pick a single version for 'tech preview', maybe we should pick the latest? That won't make it look as if we are lagging. And we'll be confronted more with the actual maintenance.

@omajid -^ my preference is to use 2.2 instead of 2.1, but you can pick.
@quintesse @ALRubinger if the multiple version support shouldn't take too long, maybe we can wait for it. This can also be a test-case for it.

**Comment by redhat-developers-launcher** [comment]
>I'm wondering if we should perhaps take a pragmatic approach and be more lenient with errors? 

Sure, I don't disagree with you in that point. My only concern, as you already mentioned is whether the returned error code represents the expected error in the tests (for example if sending an invalid request, a 4xx is returned instead of 5xx)

**Comment by gastaldi** [comment]
>I'm wondering if we should perhaps take a pragmatic approach and be more lenient with errors?

Sure, I don't disagree with you in that point. My only concern, as you already mentioned is whether the returned error code represents the expected error in the tests (for example a 404 in some runtimes may indicate that the service is not yet online). 

Anyway, let's change the tests to be more lenient and see what happens

File: catalog/generators/rest-aspnetcore/index.ts (hunk 1 above)
Comment (omajid) [reply catalog/generators/rest-aspnetcore/index.ts]: No longer needed because 1. bug was resolved 2. we are not using the centos images anymore.

**Comment by omajid** [comment]
I have switched to .NET Core 2.2. That brought in some improvements in tests. It looks like 2.2 is more idiomatic in how HTTP return codes are used :)

Only `DatabaseUpdateIllegalPayload` and `DatabaseInsertIllegalPayload` fail by returning `400` instead of `422` in the tests I ran.

**Comment by omajid** [comment]
CI is failing with:

    ERR: ./gap: line 35: dotnet: command not found

Should I open a PR for adding dotnet to the CI machines?

**Comment by gastaldi** [comment]
@omajid you need to update the `.circleci/config.yml` file to add this runtime. It runs Ubuntu last time I checked

**Comment by gastaldi** [comment]
@omajid don't worry about that, the integration tests in CI are still under development. Does it pass if you run locally?

**Comment by omajid** [comment]
> Does it pass if you run locally?

Yes, the tests pass, aside from the HTTP status code issues I mentioned in https://github.com/fabric8-launcher/launcher-creator-backend/pull/148#issuecomment-474500597.

**Comment by gastaldi** [comment]
Then this PR is ready to be merged. @quintesse is there anything you want to review before we merge this?

**Comment by gastaldi** [comment]
@omajid can you rebase this PR with the latest changes in master?

**Comment by gastaldi** [comment]
@omajid if your commit contains the following in a single line:

`TEST_ARGS: --runtimes=dotnet`

This PR will only run the dotnet tests

**Comment by omajid** [comment]
@gastaldi Done. CI shows the same behaviour I was seeing locally: `DatabaseUpdateIllegalPayload` and `DatabaseInsertIllegalPayload` return 400 instead of the expected 422.

**Comment by quintesse** [comment]
LGTM

**Comment by gastaldi** [comment]
@omajid thank you!

**Comment by quintesse** [comment]
> Only DatabaseUpdateIllegalPayload and DatabaseInsertIllegalPayload fail by returning 400 instead of 422 in the tests I ran.

@omajid I'm currently making those test more flexible so they should start passing okay real soon now

**Comment by quintesse** [comment]
@omajid for me the "HealthCheck" tests are failing with HTTP error code 404, don't you get that in your tests?

**Pull request merged**